runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java

```diff
@@ -321,6 +318,21 @@ public Instant getWatermark() {
     }
   }
 
+  /**
+   * Listener to be called when a reader is removed from {@link MicrobatchSource#readerCache}.
+   */
+  private static class ReaderCacheRemovalListener
+      implements RemovalListener<MicrobatchSource<?, ?>, BoundedReader<?>> {
+    @Override public void onRemoval(
+        RemovalNotification<MicrobatchSource<?, ?>, BoundedReader<?>> notification) {
+      try {
+        notification.getValue().close();
```

Comment:
The removed value. See [com.google.common.cache.LocalCache](https://github.com/google/guava/blob/v19.0/guava/src/com/google/common/cache/LocalCache.java#L2655)